// viem@0.1.7/dist/chains.mjs
import {
  defineBlock,
  defineChain,
  defineTransaction,
  defineTransactionReceipt,
  defineTransactionRequest
} from "./chunk-5YM5CAWB.mjs";

// src/chains.ts
import * as chains from "@wagmi/chains";
var celoFormatters = {
  block: defineBlock({
    exclude: ["difficulty", "gasLimit", "mixHash", "nonce", "uncles"],
    format: (block) => ({
      randomness: block.randomness
    })
  }),
  transaction: defineTransaction({
    format: (transaction) => ({
      feeCurrency: transaction.feeCurrency,
      gatewayFee: transaction.gatewayFee ? BigInt(transaction.gatewayFee) : null,
      gatewayFeeRecipient: transaction.gatewayFeeRecipient
    })
  }),
  transactionReceipt: defineTransactionReceipt({
    format: (transaction) => ({
      feeCurrency: transaction.feeCurrency,
      gatewayFee: transaction.gatewayFee ? BigInt(transaction.gatewayFee) : null,
      gatewayFeeRecipient: transaction.gatewayFeeRecipient
    })
  }),
  transactionRequest: defineTransactionRequest({
    format: (transactionRequest) => ({
      feeCurrency: transactionRequest.feeCurrency,
      gatewayFee: transactionRequest.gatewayFee,
      gatewayFeeRecipient: transactionRequest.gatewayFeeRecipient
    })
  })
};
var arbitrum2 = defineChain(chains.arbitrum);
var arbitrumGoerli2 = defineChain(chains.arbitrumGoerli);
var aurora2 = defineChain(chains.aurora);
var auroraGoerli = defineChain(chains.auroraTestnet);
var avalanche2 = defineChain(chains.avalanche);
var avalancheFuji2 = defineChain(chains.avalancheFuji);
var baseGoerli2 = defineChain(chains.baseGoerli);
var boba2 = defineChain(chains.boba);
var bronos2 = defineChain(chains.bronos);
var bronosTestnet2 = defineChain(chains.bronosTestnet);
var bsc2 = defineChain(chains.bsc);
var bscTestnet2 = defineChain(chains.bscTestnet);
var canto2 = defineChain(chains.canto);
var celo2 = defineChain({
  ...chains.celo,
  formatters: celoFormatters
});
var celoAlfajores2 = defineChain({
  ...chains.celoAlfajores,
  formatters: celoFormatters
});
var crossbell2 = defineChain(chains.crossbell);
var evmos2 = defineChain(chains.evmos);
var evmosTestnet2 = defineChain(chains.evmosTestnet);
var fantom2 = defineChain(chains.fantom);
var fantomTestnet2 = defineChain(chains.fantomTestnet);
var filecoin2 = defineChain(chains.filecoin);
var filecoinCalibration2 = defineChain(chains.filecoinCalibration);
var filecoinHyperspace2 = defineChain(chains.filecoinHyperspace);
var flare2 = defineChain(chains.flare);
var flareTestnet2 = defineChain(chains.flareTestnet);
var foundry2 = defineChain(chains.foundry);
var goerli2 = defineChain(chains.goerli);
var gnosis2 = defineChain(chains.gnosis);
var gnosisChiado2 = defineChain(chains.gnosisChiado);
var hardhat2 = defineChain(chains.hardhat);
var harmonyOne2 = defineChain(chains.harmonyOne);
var iotex2 = defineChain(chains.iotex);
var iotexTestnet2 = defineChain(chains.iotexTestnet);
var localhost2 = defineChain(chains.localhost);
var mainnet2 = defineChain(chains.mainnet);
var metis2 = defineChain(chains.metis);
var metisGoerli2 = defineChain(chains.metisGoerli);
var moonbaseAlpha2 = defineChain(chains.moonbaseAlpha);
var moonbeam2 = defineChain(chains.moonbeam);
var moonriver2 = defineChain(chains.moonriver);
var okc2 = defineChain(chains.okc);
var optimism2 = defineChain(chains.optimism);
var optimismGoerli2 = defineChain(chains.optimismGoerli);
var polygon2 = defineChain(chains.polygon);
var polygonMumbai2 = defineChain(chains.polygonMumbai);
var polygonZkEvmTestnet2 = defineChain(chains.polygonZkEvmTestnet);
var sepolia2 = defineChain(chains.sepolia);
var shardeumSphinx2 = defineChain(chains.shardeumSphinx);
var songbird2 = defineChain(chains.songbird);
var songbirdTestnet2 = defineChain(chains.songbirdTestnet);
var taraxa2 = defineChain(chains.taraxa);
var taraxaTestnet2 = defineChain(chains.taraxaTestnet);
var telos2 = defineChain(chains.telos);
var telosTestnet2 = defineChain(chains.telosTestnet);
var zhejiang2 = defineChain(chains.zhejiang);
var zkSync2 = defineChain(chains.zkSync);
var zkSyncTestnet2 = defineChain(chains.zkSyncTestnet);
export {
  arbitrum2 as arbitrum,
  arbitrumGoerli2 as arbitrumGoerli,
  aurora2 as aurora,
  auroraGoerli,
  avalanche2 as avalanche,
  avalancheFuji2 as avalancheFuji,
  baseGoerli2 as baseGoerli,
  boba2 as boba,
  bronos2 as bronos,
  bronosTestnet2 as bronosTestnet,
  bsc2 as bsc,
  bscTestnet2 as bscTestnet,
  canto2 as canto,
  celo2 as celo,
  celoAlfajores2 as celoAlfajores,
  crossbell2 as crossbell,
  defineChain,
  evmos2 as evmos,
  evmosTestnet2 as evmosTestnet,
  fantom2 as fantom,
  fantomTestnet2 as fantomTestnet,
  filecoin2 as filecoin,
  filecoinCalibration2 as filecoinCalibration,
  filecoinHyperspace2 as filecoinHyperspace,
  flare2 as flare,
  flareTestnet2 as flareTestnet,
  foundry2 as foundry,
  gnosis2 as gnosis,
  gnosisChiado2 as gnosisChiado,
  goerli2 as goerli,
  hardhat2 as hardhat,
  harmonyOne2 as harmonyOne,
  iotex2 as iotex,
  iotexTestnet2 as iotexTestnet,
  localhost2 as localhost,
  mainnet2 as mainnet,
  metis2 as metis,
  metisGoerli2 as metisGoerli,
  moonbaseAlpha2 as moonbaseAlpha,
  moonbeam2 as moonbeam,
  moonriver2 as moonriver,
  okc2 as okc,
  optimism2 as optimism,
  optimismGoerli2 as optimismGoerli,
  polygon2 as polygon,
  polygonMumbai2 as polygonMumbai,
  polygonZkEvmTestnet2 as polygonZkEvmTestnet,
  sepolia2 as sepolia,
  shardeumSphinx2 as shardeumSphinx,
  songbird2 as songbird,
  songbirdTestnet2 as songbirdTestnet,
  taraxa2 as taraxa,
  taraxaTestnet2 as taraxaTestnet,
  telos2 as telos,
  telosTestnet2 as telosTestnet,
  zhejiang2 as zhejiang,
  zkSync2 as zkSync,
  zkSyncTestnet2 as zkSyncTestnet
};
