// viem@0.1.3/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }




var _chunkMC6HQXR7js = require('./chunk-MC6HQXR7.js');






































































































































































































































var _chunkZ62X7UMWjs = require('./chunk-Z62X7UMW.js');

// src/clients/transports/createTransport.ts
function createTransport({
  key,
  name,
  request,
  retryCount = 3,
  retryDelay = 150,
  timeout,
  type
}, value) {
  return {
    config: { key, name, request, retryCount, retryDelay, timeout, type },
    request: _chunkZ62X7UMWjs.buildRequest.call(void 0, request, { retryCount, retryDelay }),
    value
  };
}

// src/clients/transports/custom.ts
function custom(provider, config = {}) {
  const { key = "custom", name = "Custom Provider", retryDelay } = config;
  return ({ retryCount: defaultRetryCount }) => createTransport({
    key,
    name,
    request: provider.request.bind(provider),
    retryCount: _nullishCoalesce(config.retryCount, () => ( defaultRetryCount)),
    retryDelay,
    type: "custom"
  });
}

// src/clients/transports/fallback.ts
function fallback(transports, config = {}) {
  const { key = "fallback", name = "Fallback", retryCount, retryDelay } = config;
  return ({ chain }) => createTransport(
    {
      key,
      name,
      async request({ method, params }) {
        const fetch = async (i = 0) => {
          const transport = transports[i]({ chain, retryCount: 0 });
          try {
            return await transport.request({
              method,
              params
            });
          } catch (err) {
            if (_chunkZ62X7UMWjs.isDeterministicError.call(void 0, err))
              throw err;
            if (i === transports.length - 1)
              throw err;
            return fetch(i + 1);
          }
        };
        return fetch();
      },
      retryCount,
      retryDelay,
      type: "fallback"
    },
    {
      transports: transports.map(
        (fn) => fn({ chain, retryCount: 0 })
      )
    }
  );
}

// src/clients/transports/http.ts
function http(url, config = {}) {
  const {
    fetchOptions,
    key = "http",
    name = "HTTP JSON-RPC",
    retryDelay,
    timeout = 1e4
  } = config;
  return ({ chain, retryCount: defaultRetryCount }) => {
    const retryCount = _nullishCoalesce(config.retryCount, () => ( defaultRetryCount));
    const url_ = url || _optionalChain([chain, 'optionalAccess', _ => _.rpcUrls, 'access', _2 => _2.default, 'access', _3 => _3.http, 'access', _4 => _4[0]]);
    if (!url_)
      throw new (0, _chunkZ62X7UMWjs.UrlRequiredError)();
    return createTransport(
      {
        key,
        name,
        async request({ method, params }) {
          const { result } = await _chunkZ62X7UMWjs.rpc.http(url_, {
            body: {
              method,
              params
            },
            fetchOptions,
            timeout
          });
          return result;
        },
        retryCount,
        retryDelay,
        timeout,
        type: "http"
      },
      {
        url
      }
    );
  };
}

// src/clients/transports/webSocket.ts
function webSocket(url, config = {}) {
  const {
    key = "webSocket",
    name = "WebSocket JSON-RPC",
    retryDelay,
    timeout = 1e4
  } = config;
  return ({ chain, retryCount: defaultRetryCount }) => {
    const retryCount = _nullishCoalesce(config.retryCount, () => ( defaultRetryCount));
    const url_ = url || _optionalChain([chain, 'optionalAccess', _5 => _5.rpcUrls, 'access', _6 => _6.default, 'access', _7 => _7.webSocket, 'optionalAccess', _8 => _8[0]]);
    if (!url_)
      throw new (0, _chunkZ62X7UMWjs.UrlRequiredError)();
    return createTransport(
      {
        key,
        name,
        async request({ method, params }) {
          const socket = await _chunkZ62X7UMWjs.getSocket.call(void 0, url_);
          const { result } = await _chunkZ62X7UMWjs.rpc.webSocketAsync(socket, {
            body: { method, params },
            timeout
          });
          return result;
        },
        retryCount,
        retryDelay,
        timeout,
        type: "webSocket"
      },
      {
        getSocket() {
          return _chunkZ62X7UMWjs.getSocket.call(void 0, url_);
        },
        async subscribe({ params, onData, onError }) {
          const socket = await _chunkZ62X7UMWjs.getSocket.call(void 0, url_);
          const { result: subscriptionId } = await new Promise(
            (resolve, reject) => _chunkZ62X7UMWjs.rpc.webSocket(socket, {
              body: {
                method: "eth_subscribe",
                params
              },
              onData: (data) => {
                if (typeof data.id === "number") {
                  resolve(data);
                  return;
                }
                onData(data);
              },
              onError: (error) => {
                reject(error);
                _optionalChain([onError, 'optionalCall', _9 => _9(error)]);
              }
            })
          );
          return {
            subscriptionId,
            async unsubscribe() {
              return new Promise(
                (resolve, reject) => _chunkZ62X7UMWjs.rpc.webSocket(socket, {
                  body: {
                    method: "eth_unsubscribe",
                    params: [subscriptionId]
                  },
                  onData: resolve,
                  onError: reject
                })
              );
            }
          };
        }
      }
    );
  };
}

// src/utils/uid.ts
var size2 = 256;
var index = size2;
var buffer;
function uid(length = 11) {
  if (!buffer || index + length > size2 * 2) {
    buffer = "";
    index = 0;
    for (let i = 0; i < size2; i++) {
      buffer += (256 + Math.random() * 256 | 0).toString(16).substring(1);
    }
  }
  return buffer.substring(index, index++ + length);
}

// src/clients/createClient.ts
function createClient({
  chain,
  key = "base",
  name = "Base Client",
  pollingInterval = 4e3,
  transport,
  type = "base"
}) {
  const { config, request, value } = transport({ chain });
  return {
    chain,
    key,
    name,
    pollingInterval,
    request,
    transport: { ...config, ...value },
    type,
    uid: uid()
  };
}

// src/clients/decorators/public.ts
var publicActions = (client) => ({
  call: (args) => _chunkZ62X7UMWjs.call.call(void 0, client, args),
  createBlockFilter: () => _chunkZ62X7UMWjs.createBlockFilter.call(void 0, client),
  createContractEventFilter: (args) => _chunkZ62X7UMWjs.createContractEventFilter.call(void 0, client, args),
  createEventFilter: (args) => _chunkZ62X7UMWjs.createEventFilter.call(void 0, client, args),
  createPendingTransactionFilter: () => _chunkZ62X7UMWjs.createPendingTransactionFilter.call(void 0, client),
  estimateContractGas: (args) => _chunkZ62X7UMWjs.estimateContractGas.call(void 0, client, args),
  estimateGas: (args) => _chunkZ62X7UMWjs.estimateGas.call(void 0, client, args),
  getBalance: (args) => _chunkZ62X7UMWjs.getBalance.call(void 0, client, args),
  getBlock: (args) => _chunkZ62X7UMWjs.getBlock.call(void 0, client, args),
  getBlockNumber: (args) => _chunkZ62X7UMWjs.getBlockNumber.call(void 0, client, args),
  getBlockTransactionCount: (args) => _chunkZ62X7UMWjs.getBlockTransactionCount.call(void 0, client, args),
  getBytecode: (args) => _chunkZ62X7UMWjs.getBytecode.call(void 0, client, args),
  getChainId: () => _chunkZ62X7UMWjs.getChainId.call(void 0, client),
  getEnsAddress: (args) => _chunkMC6HQXR7js.getEnsAddress.call(void 0, client, args),
  getEnsName: (args) => _chunkMC6HQXR7js.getEnsName.call(void 0, client, args),
  getFeeHistory: (args) => _chunkZ62X7UMWjs.getFeeHistory.call(void 0, client, args),
  getFilterChanges: (args) => _chunkZ62X7UMWjs.getFilterChanges.call(void 0, client, args),
  getFilterLogs: (args) => _chunkZ62X7UMWjs.getFilterLogs.call(void 0, client, args),
  getGasPrice: () => _chunkZ62X7UMWjs.getGasPrice.call(void 0, client),
  getLogs: (args) => _chunkZ62X7UMWjs.getLogs.call(void 0, client, args),
  getStorageAt: (args) => _chunkZ62X7UMWjs.getStorageAt.call(void 0, client, args),
  getTransaction: (args) => _chunkZ62X7UMWjs.getTransaction.call(void 0, client, args),
  getTransactionConfirmations: (args) => _chunkZ62X7UMWjs.getTransactionConfirmations.call(void 0, client, args),
  getTransactionReceipt: (args) => _chunkZ62X7UMWjs.getTransactionReceipt.call(void 0, client, args),
  multicall: (args) => _chunkZ62X7UMWjs.multicall.call(void 0, client, args),
  readContract: (args) => _chunkZ62X7UMWjs.readContract.call(void 0, client, args),
  simulateContract: (args) => _chunkZ62X7UMWjs.simulateContract.call(void 0, client, args),
  uninstallFilter: (args) => _chunkZ62X7UMWjs.uninstallFilter.call(void 0, client, args),
  waitForTransactionReceipt: (args) => _chunkZ62X7UMWjs.waitForTransactionReceipt.call(void 0, client, args),
  watchBlocks: (args) => _chunkZ62X7UMWjs.watchBlocks.call(void 0, client, args),
  watchBlockNumber: (args) => _chunkZ62X7UMWjs.watchBlockNumber.call(void 0, client, args),
  watchContractEvent: (args) => _chunkZ62X7UMWjs.watchContractEvent.call(void 0, client, args),
  watchEvent: (args) => _chunkZ62X7UMWjs.watchEvent.call(void 0, client, args),
  watchPendingTransactions: (args) => _chunkZ62X7UMWjs.watchPendingTransactions.call(void 0, client, args)
});

// src/clients/decorators/test.ts
var testActions = (client) => ({
  dropTransaction: (args) => _chunkZ62X7UMWjs.dropTransaction.call(void 0, client, args),
  getAutomine: () => _chunkZ62X7UMWjs.getAutomine.call(void 0, client),
  getTxpoolContent: () => _chunkZ62X7UMWjs.getTxpoolContent.call(void 0, client),
  getTxpoolStatus: () => _chunkZ62X7UMWjs.getTxpoolStatus.call(void 0, client),
  impersonateAccount: (args) => _chunkZ62X7UMWjs.impersonateAccount.call(void 0, client, args),
  increaseTime: (args) => _chunkZ62X7UMWjs.increaseTime.call(void 0, client, args),
  inspectTxpool: () => _chunkZ62X7UMWjs.inspectTxpool.call(void 0, client),
  mine: (args) => _chunkZ62X7UMWjs.mine.call(void 0, client, args),
  removeBlockTimestampInterval: () => _chunkZ62X7UMWjs.removeBlockTimestampInterval.call(void 0, client),
  reset: (args) => _chunkZ62X7UMWjs.reset.call(void 0, client, args),
  revert: (args) => _chunkZ62X7UMWjs.revert.call(void 0, client, args),
  sendUnsignedTransaction: (args) => _chunkZ62X7UMWjs.sendUnsignedTransaction.call(void 0, client, args),
  setAutomine: (args) => _chunkZ62X7UMWjs.setAutomine.call(void 0, client, args),
  setBalance: (args) => _chunkZ62X7UMWjs.setBalance.call(void 0, client, args),
  setBlockGasLimit: (args) => _chunkZ62X7UMWjs.setBlockGasLimit.call(void 0, client, args),
  setBlockTimestampInterval: (args) => _chunkZ62X7UMWjs.setBlockTimestampInterval.call(void 0, client, args),
  setCode: (args) => _chunkZ62X7UMWjs.setCode.call(void 0, client, args),
  setCoinbase: (args) => _chunkZ62X7UMWjs.setCoinbase.call(void 0, client, args),
  setIntervalMining: (args) => _chunkZ62X7UMWjs.setIntervalMining.call(void 0, client, args),
  setLoggingEnabled: (args) => _chunkZ62X7UMWjs.setLoggingEnabled.call(void 0, client, args),
  setMinGasPrice: (args) => _chunkZ62X7UMWjs.setMinGasPrice.call(void 0, client, args),
  setNextBlockBaseFeePerGas: (args) => _chunkZ62X7UMWjs.setNextBlockBaseFeePerGas.call(void 0, client, args),
  setNextBlockTimestamp: (args) => _chunkZ62X7UMWjs.setNextBlockTimestamp.call(void 0, client, args),
  setNonce: (args) => _chunkZ62X7UMWjs.setNonce.call(void 0, client, args),
  setRpcUrl: (args) => _chunkZ62X7UMWjs.setRpcUrl.call(void 0, client, args),
  setStorageAt: (args) => _chunkZ62X7UMWjs.setStorageAt.call(void 0, client, args),
  snapshot: () => _chunkZ62X7UMWjs.snapshot.call(void 0, client),
  stopImpersonatingAccount: (args) => _chunkZ62X7UMWjs.stopImpersonatingAccount.call(void 0, client, args)
});

// src/clients/decorators/wallet.ts
var walletActions = (client) => ({
  addChain: (args) => _chunkZ62X7UMWjs.addChain.call(void 0, client, args),
  deployContract: (args) => _chunkZ62X7UMWjs.deployContract.call(void 0, client, args),
  getAddresses: () => _chunkZ62X7UMWjs.getAddresses.call(void 0, client),
  getChainId: () => _chunkZ62X7UMWjs.getChainId.call(void 0, client),
  getPermissions: () => _chunkZ62X7UMWjs.getPermissions.call(void 0, client),
  requestAddresses: () => _chunkZ62X7UMWjs.requestAddresses.call(void 0, client),
  requestPermissions: (args) => _chunkZ62X7UMWjs.requestPermissions.call(void 0, client, args),
  sendTransaction: (args) => _chunkZ62X7UMWjs.sendTransaction.call(void 0, client, args),
  signMessage: (args) => _chunkZ62X7UMWjs.signMessage.call(void 0, client, args),
  switchChain: (args) => _chunkZ62X7UMWjs.switchChain.call(void 0, client, args),
  watchAsset: (args) => _chunkZ62X7UMWjs.watchAsset.call(void 0, client, args),
  writeContract: (args) => _chunkZ62X7UMWjs.writeContract.call(void 0, client, args)
});

// src/clients/createPublicClient.ts
function createPublicClient({
  chain,
  key = "public",
  name = "Public Client",
  transport,
  pollingInterval
}) {
  const client = createClient({
    chain,
    key,
    name,
    pollingInterval,
    transport,
    type: "publicClient"
  });
  return {
    ...client,
    ...publicActions(client)
  };
}

// src/clients/createTestClient.ts
function createTestClient({
  chain,
  key = "test",
  name = "Test Client",
  mode,
  pollingInterval,
  transport
}) {
  const client = {
    ...createClient({
      chain,
      key,
      name,
      pollingInterval,
      transport,
      type: "testClient"
    }),
    mode
  };
  return {
    ...client,
    ...testActions(client)
  };
}

// src/clients/createWalletClient.ts
function createWalletClient({
  transport,
  key = "wallet",
  name = "Wallet Client",
  pollingInterval
}) {
  const client = createClient({
    key,
    name,
    pollingInterval,
    transport: () => transport({ retryCount: 0 }),
    type: "walletClient"
  });
  return {
    ...client,
    ...walletActions(client)
  };
}






































































































































































exports.AbiConstructorNotFoundError = _chunkZ62X7UMWjs.AbiConstructorNotFoundError; exports.AbiConstructorParamsNotFoundError = _chunkZ62X7UMWjs.AbiConstructorParamsNotFoundError; exports.AbiDecodingDataSizeInvalidError = _chunkZ62X7UMWjs.AbiDecodingDataSizeInvalidError; exports.AbiDecodingZeroDataError = _chunkZ62X7UMWjs.AbiDecodingZeroDataError; exports.AbiEncodingArrayLengthMismatchError = _chunkZ62X7UMWjs.AbiEncodingArrayLengthMismatchError; exports.AbiEncodingLengthMismatchError = _chunkZ62X7UMWjs.AbiEncodingLengthMismatchError; exports.AbiErrorInputsNotFoundError = _chunkZ62X7UMWjs.AbiErrorInputsNotFoundError; exports.AbiErrorNotFoundError = _chunkZ62X7UMWjs.AbiErrorNotFoundError; exports.AbiErrorSignatureNotFoundError = _chunkZ62X7UMWjs.AbiErrorSignatureNotFoundError; exports.AbiEventNotFoundError = _chunkZ62X7UMWjs.AbiEventNotFoundError; exports.AbiEventSignatureNotFoundError = _chunkZ62X7UMWjs.AbiEventSignatureNotFoundError; exports.AbiFunctionNotFoundError = _chunkZ62X7UMWjs.AbiFunctionNotFoundError; exports.AbiFunctionOutputsNotFoundError = _chunkZ62X7UMWjs.AbiFunctionOutputsNotFoundError; exports.AbiFunctionSignatureNotFoundError = _chunkZ62X7UMWjs.AbiFunctionSignatureNotFoundError; exports.BaseError = _chunkZ62X7UMWjs.BaseError; exports.BlockNotFoundError = _chunkZ62X7UMWjs.BlockNotFoundError; exports.CallExecutionError = _chunkZ62X7UMWjs.CallExecutionError; exports.ChainDoesNotSupportContract = _chunkZ62X7UMWjs.ChainDoesNotSupportContract; exports.ContractFunctionExecutionError = _chunkZ62X7UMWjs.ContractFunctionExecutionError; exports.ContractFunctionRevertedError = _chunkZ62X7UMWjs.ContractFunctionRevertedError; exports.ContractFunctionZeroDataError = _chunkZ62X7UMWjs.ContractFunctionZeroDataError; exports.DataLengthTooLongError = _chunkZ62X7UMWjs.DataLengthTooLongError; exports.DataLengthTooShortError = _chunkZ62X7UMWjs.DataLengthTooShortError; exports.EstimateGasExecutionError = _chunkZ62X7UMWjs.EstimateGasExecutionError; exports.ExecutionRevertedError = _chunkZ62X7UMWjs.ExecutionRevertedError; exports.FeeCapTooHighError = _chunkZ62X7UMWjs.FeeCapTooHighError; exports.FeeCapTooLowError = _chunkZ62X7UMWjs.FeeCapTooLowError; exports.FilterTypeNotSupportedError = _chunkZ62X7UMWjs.FilterTypeNotSupportedError; exports.HttpRequestError = _chunkZ62X7UMWjs.HttpRequestError; exports.InsufficientFundsError = _chunkZ62X7UMWjs.InsufficientFundsError; exports.InternalRpcError = _chunkZ62X7UMWjs.InternalRpcError; exports.IntrinsicGasTooHighError = _chunkZ62X7UMWjs.IntrinsicGasTooHighError; exports.IntrinsicGasTooLowError = _chunkZ62X7UMWjs.IntrinsicGasTooLowError; exports.InvalidAbiDecodingTypeError = _chunkZ62X7UMWjs.InvalidAbiDecodingTypeError; exports.InvalidAbiEncodingTypeError = _chunkZ62X7UMWjs.InvalidAbiEncodingTypeError; exports.InvalidAddressError = _chunkZ62X7UMWjs.InvalidAddressError; exports.InvalidArrayError = _chunkZ62X7UMWjs.InvalidArrayError; exports.InvalidBytesBooleanError = _chunkZ62X7UMWjs.InvalidBytesBooleanError; exports.InvalidDefinitionTypeError = _chunkZ62X7UMWjs.InvalidDefinitionTypeError; exports.InvalidHexBooleanError = _chunkZ62X7UMWjs.InvalidHexBooleanError; exports.InvalidHexValueError = _chunkZ62X7UMWjs.InvalidHexValueError; exports.InvalidInputRpcError = _chunkZ62X7UMWjs.InvalidInputRpcError; exports.InvalidParamsRpcError = _chunkZ62X7UMWjs.InvalidParamsRpcError; exports.InvalidRequestRpcError = _chunkZ62X7UMWjs.InvalidRequestRpcError; exports.JsonRpcVersionUnsupportedError = _chunkZ62X7UMWjs.JsonRpcVersionUnsupportedError; exports.LimitExceededRpcError = _chunkZ62X7UMWjs.LimitExceededRpcError; exports.MethodNotFoundRpcError = _chunkZ62X7UMWjs.MethodNotFoundRpcError; exports.MethodNotSupportedRpcError = _chunkZ62X7UMWjs.MethodNotSupportedRpcError; exports.NonceMaxValueError = _chunkZ62X7UMWjs.NonceMaxValueError; exports.NonceTooHighError = _chunkZ62X7UMWjs.NonceTooHighError; exports.NonceTooLowError = _chunkZ62X7UMWjs.NonceTooLowError; exports.OffsetOutOfBoundsError = _chunkZ62X7UMWjs.OffsetOutOfBoundsError; exports.ParseRpcError = _chunkZ62X7UMWjs.ParseRpcError; exports.RawContractError = _chunkZ62X7UMWjs.RawContractError; exports.RequestError = _chunkZ62X7UMWjs.RequestError; exports.ResourceNotFoundRpcError = _chunkZ62X7UMWjs.ResourceNotFoundRpcError; exports.ResourceUnavailableRpcError = _chunkZ62X7UMWjs.ResourceUnavailableRpcError; exports.RpcError = _chunkZ62X7UMWjs.RpcError; exports.RpcRequestError = _chunkZ62X7UMWjs.RpcRequestError; exports.SizeExceedsPaddingSizeError = _chunkZ62X7UMWjs.SizeExceedsPaddingSizeError; exports.SwitchChainError = _chunkZ62X7UMWjs.SwitchChainError; exports.TimeoutError = _chunkZ62X7UMWjs.TimeoutError; exports.TipAboveFeeCapError = _chunkZ62X7UMWjs.TipAboveFeeCapError; exports.TransactionExecutionError = _chunkZ62X7UMWjs.TransactionExecutionError; exports.TransactionNotFoundError = _chunkZ62X7UMWjs.TransactionNotFoundError; exports.TransactionReceiptNotFoundError = _chunkZ62X7UMWjs.TransactionReceiptNotFoundError; exports.TransactionRejectedRpcError = _chunkZ62X7UMWjs.TransactionRejectedRpcError; exports.TransactionTypeNotSupportedError = _chunkZ62X7UMWjs.TransactionTypeNotSupportedError; exports.UnknownNodeError = _chunkZ62X7UMWjs.UnknownNodeError; exports.UnknownRpcError = _chunkZ62X7UMWjs.UnknownRpcError; exports.UrlRequiredError = _chunkZ62X7UMWjs.UrlRequiredError; exports.UserRejectedRequestError = _chunkZ62X7UMWjs.UserRejectedRequestError; exports.WaitForTransactionReceiptTimeoutError = _chunkZ62X7UMWjs.WaitForTransactionReceiptTimeoutError; exports.WebSocketRequestError = _chunkZ62X7UMWjs.WebSocketRequestError; exports.boolToBytes = _chunkZ62X7UMWjs.boolToBytes; exports.boolToHex = _chunkZ62X7UMWjs.boolToHex; exports.bytesToBigint = _chunkZ62X7UMWjs.bytesToBigint; exports.bytesToBool = _chunkZ62X7UMWjs.bytesToBool; exports.bytesToHex = _chunkZ62X7UMWjs.bytesToHex; exports.bytesToNumber = _chunkZ62X7UMWjs.bytesToNumber; exports.bytesToString = _chunkZ62X7UMWjs.bytesToString; exports.createClient = createClient; exports.createPublicClient = createPublicClient; exports.createTestClient = createTestClient; exports.createTransport = createTransport; exports.createWalletClient = createWalletClient; exports.custom = custom; exports.decodeAbiParameters = _chunkZ62X7UMWjs.decodeAbiParameters; exports.decodeErrorResult = _chunkZ62X7UMWjs.decodeErrorResult; exports.decodeEventLog = _chunkZ62X7UMWjs.decodeEventLog; exports.decodeFunctionData = _chunkZ62X7UMWjs.decodeFunctionData; exports.decodeFunctionResult = _chunkZ62X7UMWjs.decodeFunctionResult; exports.defineBlock = _chunkZ62X7UMWjs.defineBlock; exports.defineChain = _chunkZ62X7UMWjs.defineChain; exports.defineTransaction = _chunkZ62X7UMWjs.defineTransaction; exports.defineTransactionReceipt = _chunkZ62X7UMWjs.defineTransactionReceipt; exports.defineTransactionRequest = _chunkZ62X7UMWjs.defineTransactionRequest; exports.encodeAbiParameters = _chunkZ62X7UMWjs.encodeAbiParameters; exports.encodeDeployData = _chunkZ62X7UMWjs.encodeDeployData; exports.encodeErrorResult = _chunkZ62X7UMWjs.encodeErrorResult; exports.encodeEventTopics = _chunkZ62X7UMWjs.encodeEventTopics; exports.encodeFunctionData = _chunkZ62X7UMWjs.encodeFunctionData; exports.encodeFunctionResult = _chunkZ62X7UMWjs.encodeFunctionResult; exports.encodePacked = _chunkZ62X7UMWjs.encodePacked; exports.etherUnits = _chunkZ62X7UMWjs.etherUnits; exports.fallback = fallback; exports.formatBlock = _chunkZ62X7UMWjs.formatBlock; exports.formatEther = _chunkZ62X7UMWjs.formatEther; exports.formatGwei = _chunkZ62X7UMWjs.formatGwei; exports.formatTransaction = _chunkZ62X7UMWjs.formatTransaction; exports.formatTransactionRequest = _chunkZ62X7UMWjs.formatTransactionRequest; exports.formatUnits = _chunkZ62X7UMWjs.formatUnits; exports.fromBytes = _chunkZ62X7UMWjs.fromBytes; exports.fromHex = _chunkZ62X7UMWjs.fromHex; exports.fromRlp = _chunkZ62X7UMWjs.fromRlp; exports.getAbiItem = _chunkZ62X7UMWjs.getAbiItem; exports.getAccount = _chunkZ62X7UMWjs.getAccount; exports.getAddress = _chunkZ62X7UMWjs.getAddress; exports.getContractAddress = _chunkZ62X7UMWjs.getContractAddress; exports.getContractError = _chunkZ62X7UMWjs.getContractError; exports.getCreate2Address = _chunkZ62X7UMWjs.getCreate2Address; exports.getCreateAddress = _chunkZ62X7UMWjs.getCreateAddress; exports.getEventSelector = _chunkZ62X7UMWjs.getEventSelector; exports.getFunctionSelector = _chunkZ62X7UMWjs.getFunctionSelector; exports.gweiUnits = _chunkZ62X7UMWjs.gweiUnits; exports.hexToBigInt = _chunkZ62X7UMWjs.hexToBigInt; exports.hexToBool = _chunkZ62X7UMWjs.hexToBool; exports.hexToBytes = _chunkZ62X7UMWjs.hexToBytes; exports.hexToNumber = _chunkZ62X7UMWjs.hexToNumber; exports.hexToString = _chunkZ62X7UMWjs.hexToString; exports.http = http; exports.isAddress = _chunkZ62X7UMWjs.isAddress; exports.isAddressEqual = _chunkZ62X7UMWjs.isAddressEqual; exports.isBytes = _chunkZ62X7UMWjs.isBytes; exports.isHex = _chunkZ62X7UMWjs.isHex; exports.keccak256 = _chunkZ62X7UMWjs.keccak256; exports.labelhash = _chunkMC6HQXR7js.labelhash; exports.multicall3Abi = _chunkZ62X7UMWjs.multicall3Abi; exports.namehash = _chunkMC6HQXR7js.namehash; exports.numberToBytes = _chunkZ62X7UMWjs.numberToBytes; exports.numberToHex = _chunkZ62X7UMWjs.numberToHex; exports.pad = _chunkZ62X7UMWjs.pad; exports.padBytes = _chunkZ62X7UMWjs.padBytes; exports.padHex = _chunkZ62X7UMWjs.padHex; exports.parseAbi = _chunkZ62X7UMWjs.parseAbi; exports.parseAbiItem = _chunkZ62X7UMWjs.parseAbiItem; exports.parseAbiParameter = _chunkZ62X7UMWjs.parseAbiParameter; exports.parseAbiParameters = _chunkZ62X7UMWjs.parseAbiParameters; exports.parseEther = _chunkZ62X7UMWjs.parseEther; exports.parseGwei = _chunkZ62X7UMWjs.parseGwei; exports.parseUnits = _chunkZ62X7UMWjs.parseUnits; exports.size = _chunkZ62X7UMWjs.size; exports.slice = _chunkZ62X7UMWjs.slice; exports.sliceBytes = _chunkZ62X7UMWjs.sliceBytes; exports.sliceHex = _chunkZ62X7UMWjs.sliceHex; exports.stringToBytes = _chunkZ62X7UMWjs.stringToBytes; exports.stringToHex = _chunkZ62X7UMWjs.stringToHex; exports.stringify = _chunkZ62X7UMWjs.stringify; exports.toBytes = _chunkZ62X7UMWjs.toBytes; exports.toHex = _chunkZ62X7UMWjs.toHex; exports.toRlp = _chunkZ62X7UMWjs.toRlp; exports.transactionType = _chunkZ62X7UMWjs.transactionType; exports.trim = _chunkZ62X7UMWjs.trim; exports.webSocket = webSocket; exports.weiUnits = _chunkZ62X7UMWjs.weiUnits;
